20251028-BRS-workflow-versioning.md

```diff
@@ -0,0 +1,416 @@
+is without registering this in any way with the runtime ahead of time.
+
+That's fantastic for changes both large and small, but it means that there's some regular clerical work involved to
+address even the smallest of changes (e.g., flip an inadvertent negative sign somewhere in the code).
+
+Neither solution can address a bug that's been introduced to a workflow and causing it to crash (see the Caveat
+section towards the end of this document). However, we can certainly accommodate the desire to apply minimalistic changes
+to a workflow while opting out of a more-involved process.
+
+In my version of the patching proposal, it's providing a named "Band-Aid" for code, wrapping the "old" and "new" code in 
+a simple `if/then` statement with a named identifier. Patches can be applied throughout the whole of a workflow
+subject to the following constraints:
+- Patches cannot overlap one another
+- While patches are applied within a version, they are not themselves versioned
+- Patches must still be deterministically replayable - in other words, even if they introduce a new path, if it was not
+evaluated during the original run (prior to replays), it will not be evaluated during replays as this would violate
+our deterministic constraint.
+
+Generally, the standing guidance would be that developers simply create new types to version their code. As a
+(perfectly valid) exception, one or more patches should be applied. They should be understood to fix minor things in the
+workflow and are not intended to be a replacement for the whole-type versioning, but rather a complementary and necessary
+mechanism to supplement whole-type versioning in limited conditions.
+
+## Assumptions
+- All workflow base types and activities are registered in the Workflow SDK at startup as is necessary today
+- Versioning should only be performed on workflows that are not in a re-run state - otherwise, the routing mechanism in the
+SDK will be responsible for ensuring that the same versioned workflow type is used as was last re-run, absent patch changes.
+- This proposal addresses workflow versioning only and not activity versioning. Activities are not required to be
+deterministic and can be swapped out, although this generally isn't recommended. Clear documentation should generally
+advise against sharing activities across workflows due to the potential for unexpected ramifications.
+- All examples in this workflow are written using C# and are subject to differences across SDK implementation as makes 
+sense in each example given language norms, conventions and available tooling.
+- This proposal assumes a monotonically increasing version number applied to the suffix of the workflow type as one 
+of many potentially configurable versioning schemes. Despite opting to use this for consistency and clarity throughout
+the proposal, I'm certainly not trying to give the impression that it's the only scheme any Workflow SDK could implement.
```

Comment:
Workflows are registered by a name, not strictly type. Registration by type is an SDK helper feature- at he data level they are routed and referenced by string.

```suggestion
- This proposal assumes a monotonically increasing version number applied to the suffix of the workflow name as one 
of many potentially configurable versioning schemes. Despite opting to use this for consistency and clarity throughout
the proposal, I'm certainly not trying to give the impression that it's the only scheme any Workflow SDK could implement.
```